File: go/0687_longest_univalue_path.go
package main

func longestUnivaluePath(root *TreeNode) int {
	if root == nil {
		return 0
	}
	res := 0
	doLongestUnivaluePath(root, &res)
	return res
}

func doLongestUnivaluePath(node *TreeNode, res *int) int {
	switch {
	case node.Left == nil && node.Right == nil:
		return 0
	case node.Left != nil && node.Right == nil:
		h := doLongestUnivaluePath(node.Left, res)
		if node.Val == node.Left.Val {
			*res = max(*res, h+1)
			return h + 1
		}
		return 0
	case node.Left == nil && node.Right != nil:
		h := doLongestUnivaluePath(node.Right, res)
		if node.Val == node.Right.Val {
			*res = max(*res, h+1)
			return h + 1
		}
		return 0
	default:
		hl := doLongestUnivaluePath(node.Left, res)
		hr := doLongestUnivaluePath(node.Right, res)
		switch {
		case node.Val == node.Left.Val && node.Val == node.Right.Val:
			*res = max(*res, hl+hr+2)
			return max(hl, hr) + 1
		case node.Val != node.Left.Val && node.Val == node.Right.Val:
			*res = max(*res, hr+1)
			return hr + 1
		case node.Val == node.Left.Val && node.Val != node.Right.Val:
			*res = max(*res, hl+1)
			return hl + 1
		default:
			return 0
		}
	}
}

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}

File: python/0687_longest_univalue_path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        if not root:
            return 0

        res = [0]
        leftpath = rightpath = 0
        if root.left:
            if root.val == root.left.val:
                leftpath = 1 + self.helper(root.left, root.val, 1, res)
            else:
                self.helper(root.left, root.val, 0, res)
        if root.right:
            if root.val == root.right.val:
                rightpath = 1 + self.helper(root.right, root.val, 1, res)
            else:
                self.helper(root.right, root.val, 0, res)
        return max(res[0], leftpath + rightpath)

    def helper(self, node, parentVal, cur, res):
        leftpath = rightpath = 0
        if node.left:
            if node.val == node.left.val:
                leftpath = 1 + self.helper(node.left, node.val, cur + 1, res)
            else:
                res[0] = max(res[0], cur)
                self.helper(node.left, node.val, 0, res)
        if node.right:
            if node.val == node.right.val:
                rightpath = 1 + self.helper(node.right, node.val, cur + 1, res)
            else:
                res[0] = max(res[0], cur)
                self.helper(node.right, node.val, 0, res)
        res[0] = max(res[0], leftpath + rightpath)
        return max(leftpath, rightpath)




